src/lib/formats/rpk.cpp

```diff
@@ -0,0 +1,459 @@
+		</resources>
+		<configuration>
+			<pcb type="standard">
+				<socket id="grom_socket" uses="gromimage"/>
+			</pcb>
+		</configuration>
+	</romset>
+
+DTD:
+	<!ELEMENT romset (resources, configuration)>
+	<!ELEMENT resources (rom|ram)+>
+	<!ELEMENT rom EMPTY>
+	<!ELEMENT ram EMPTY>
+	<!ELEMENT configuration (pcb)>
+	<!ELEMENT pcb (socket)+>
+	<!ELEMENT socket EMPTY>
+	<!ATTLIST romset version CDATA #IMPLIED>
+	<!ATTLIST rom id ID #REQUIRED
+	<!ATTLIST rom file CDATA #REQUIRED>
+	<!ATTLIST rom crc CDATA #IMPLIED>
+	<!ATTLIST rom sha1 CDATA #IMPLIED>
+	<!ATTLIST ram id ID #REQUIRED>
+	<!ATTLIST ram type (volatile|persistent) #IMPLIED>
+	<!ATTLIST ram store (internal|external) #IMPLIED>
+	<!ATTLIST ram file CDATA #IMPLIED>
+	<!ATTLIST ram length CDATA #REQUIRED>
+	<!ATTLIST pcb type CDATA #REQUIRED>
+	<!ATTLIST socket id ID #REQUIRED>
+	<!ATTLIST socket uses IDREF #REQUIRED>
+
+***************************************************************************/
+
+#include "rpk.h"
+#include "unzip.h"
+#include "xmlfile.h"
+
+#include <vector>
+
+
+/***************************************************************************
+	TYPE DEFINITIONS
+***************************************************************************/
+
+namespace
+{
+	class rpk_category_impl : public std::error_category
+	{
+	public:
+		virtual char const *name() const noexcept override { return "rpk"; }
+		virtual std::string message(int condition) const override;
+	};
+};
+
+
+/***************************************************************************
+	GLOBAL VARIABLES
+***************************************************************************/
+
+namespace
+{
+	rpk_category_impl const f_rpk_category_instance;
+};
+
+
+/***************************************************************************
+	RPK READER
+***************************************************************************/
+
+//-------------------------------------------------
+//  ctor
+//-------------------------------------------------
+
+util::rpk_reader::rpk_reader(const char **pcb_types, bool supports_ram)
+	: m_pcb_types(pcb_types)
+	, m_supports_ram(supports_ram)
+{
+}
+
+
+//-------------------------------------------------
+//  read
+//-------------------------------------------------
+
+std::error_condition util::rpk_reader::read(std::unique_ptr<random_read> &&stream, rpk_file::ptr &result) const
+{
+	// open the RPK (as a zip file)
+	util::archive_file::ptr zipfile;
+	std::error_condition ziperr = util::archive_file::open_zip(std::move(stream), zipfile);
+	if (ziperr)
+		return error::NOT_ZIP_FORMAT;
+
+	// open the layout XML
+	if (zipfile->search("layout.xml", false) < 0)
+		return error::MISSING_LAYOUT;
+
+	// prepare a buffer for the layout XML
+	std::vector<char> layout_xml_text;
+	layout_xml_text.resize(zipfile->current_uncompressed_length() + 1);
+
+	// and decompress it
+	ziperr = zipfile->decompress(&layout_xml_text[0], zipfile->current_uncompressed_length());
```

Comment:
Given how many times you use it, you should save `zipfile->current_uncompressed_length()` in a local to avoid having to go through the virtual method call every time (since archive file is an interface, the compiler can’t optimise out multiple calls as it doesn’t know if the call depends on hidden state).